src/main/java/io/github/lottetreg/httpserver/Reader.java

```diff
@@ -1,14 +1,27 @@
 package io.github.lottetreg.httpserver;
 
-import java.io.BufferedReader;
-import java.io.InputStreamReader;
+import java.util.List;
 
 public class Reader {
-  public String read(Connectionable connection) {
+  public HTTPRequest read(Connectionable connection) {
     try {
-      InputStreamReader inputStreamReader = new InputStreamReader(connection.getInputStream());
-      BufferedReader bufferedReader = new BufferedReader(inputStreamReader);
-      return bufferedReader.readLine();
+      StreamReader streamReader =
+              new StreamReader(connection.getInputStream());
+
+      String initialLine = streamReader.readLine();
+      List<String> headerLines = streamReader.readLinesUntilEmptyLine();
+
+      HTTPRequest request = new HTTPRequest(
+              new HTTPInitialLine(initialLine),
+              new HTTPHeaders(headerLines));
+
```

Comment:
Something to consider: If an invalid HTTP request is made these objects are initialized then immediately throw an exception which crashes the server. Something about the parsing/validation logic happening inside a request object (or request-related object) feels weird to me. It seems like the parsing/validating should all happen at a higher level before a request object is made. This might become more of a pain point when you need to handle `404`s so you don't need to address it now just something to think about.

Reply:
I like this idea! I had been hesitant to parse through the request twice, but if I had some object that could parse the request, then I could just use that twice when I validate and then digest the request. The acceptance tests don't account for `400`s, so we'll probably want to add a story for that in a future epic and I'll make the change then! Thank you!